crates/tx5/src/ep.rs

```diff
@@ -0,0 +1,288 @@
+    /// You probably only want to call this once.
+    pub fn listen(&self, sig_url: SigUrl) {
+        let _ = self.inner.lock().unwrap().assert_sig(sig_url, true);
+    }
+
+    /// Request that the peer connection identified by the given `peer_url`
+    /// is closed.
+    pub fn close(&self, peer_url: &PeerUrl) {
+        self.inner.lock().unwrap().drop_peer_url(peer_url);
+    }
+
+    /// Send data to a remote on this tx5 endpoint.
+    /// The future returned from this method will resolve when
+    /// the data is handed off to our networking backend.
+    pub async fn send(&self, peer_url: PeerUrl, data: Vec<u8>) -> Result<()> {
+        tokio::time::timeout(self.config.timeout, async {
+            let peer = self.inner.lock().unwrap().connect_peer(peer_url);
+            peer.ready().await;
+            peer.send(data).await
+        })
+        .await?
+    }
+
+    /// Broadcast data to all connections that happen to be open.
+    /// If no connections are open, no data will be broadcast.
+    /// The future returned from this method will resolve when all
+    /// broadcast messages have been handed off to our networking backend
+    /// (or have timed out).
+    pub async fn broadcast(&self, data: &[u8]) {
+        let peers = self.inner.lock().unwrap().clone_peers();
+
+        let timeout = self.config.timeout;
+
+        let all = peers
+            .into_iter()
+            .map(|peer| {
+                let data = data.to_vec();
+                tokio::task::spawn(async move {
+                    let _ = tokio::time::timeout(timeout, async {
+                        peer.ready().await;
+                        let _ = peer.send(data).await;
+                    })
+                    .await;
+                })
+            })
+            .collect::<Vec<_>>();
+
+        for task in all {
+            let _ = task.await;
+        }
+    }
+
+    /// Get a list of listening addresses (PeerUrls) at which this endpoint
+    /// is currently reachable.
+    pub fn get_listening_addresses(&self) -> Vec<PeerUrl> {
+        let all_sigs = self
+            .inner
+            .lock()
+            .unwrap()
+            .sig_map
+            .values()
+            .cloned()
+            .collect::<Vec<_>>();
+
+        all_sigs
+            .into_iter()
+            .filter_map(|sig| {
+                if !sig.listener {
+                    return None;
+                }
+                sig.get_peer_url()
+            })
+            .collect()
+    }
+
+    /// Get stats.
+    pub fn get_stats(&self) -> stats::Stats {
+        #[cfg(feature = "backend-go-pion")]
+        let backend = stats::StatsBackend::BackendGoPion;
+        #[cfg(feature = "backend-webrtc-rs")]
+        let backend = stats::StatsBackend::BackendWebrtcRs;
+
+        stats::Stats {
+            backend,
+            peer_url_list: self.get_listening_addresses(),
+            connection_list: Vec::new(),
```

Comment:
Note, this just returns an empty array for now. I'll flesh out the stats once we have the webrtc connections working.